File: fixtures/programming/julia/scheduler.jl
"""
Scheduling primitives for the task pipeline.

Builds on `TaskStore` and adds priority ordering and retry policies.
"""
module Scheduler

using Base.Threads
using Dates: DateTime, now
import Printf

export Priority, RetryPolicy, schedule, next_slot, @timed_run

"""Supertype of every scheduling priority."""
abstract type Priority end

struct Low <: Priority end
struct Normal <: Priority end
struct High <: Priority end

const DEFAULT_BACKOFF = 1.5

"""
    RetryPolicy(limit, backoff)

Retry behaviour attached to a scheduled task.
"""
mutable struct RetryPolicy
    limit::Int
    backoff::Float64
    attempts::Int
    RetryPolicy(limit, backoff = DEFAULT_BACKOFF) = new(limit, backoff, 0)
end

"""A slot is a millisecond offset from the epoch."""
struct Slot{T<:Integer}
    value::T
end

"""Numeric weight of a priority."""
weight(::Low) = 0
weight(::Normal) = 1
weight(::High) = 2

"""
Order tasks so that high priority names come first.

Ties keep their original relative order, because `sort` is stable.
"""
function schedule(tasks::Vector{Tuple{String,Priority}})::Vector{String}
    ordered = sort(tasks; by = t -> -weight(t[2]))
    return [name for (name, _) in ordered]
end

"""Compute the slot following `s`, spaced by the policy backoff."""
function next_slot(policy::RetryPolicy, s::Slot{T}) where {T<:Integer}
    step = ceil(T, policy.backoff * 1000)
    return Slot{T}(s.value + max(one(T), step))
end

"""Render a policy for a status report."""
function describe(p::RetryPolicy)
    return Printf.@sprintf("retry up to %d times, backoff %.2f", p.limit, p.backoff)
end

"""Time a block and report how long it took."""
macro timed_run(expr)
    return quote
        local t0 = time_ns()
        local result = $(esc(expr))
        (result, time_ns() - t0)
    end
end

end # module Scheduler

File: fixtures/programming/haskell/Scheduler.hs
{-# LANGUAGE ScopedTypeVariables #-}

-- | Scheduling primitives for the task pipeline.
--
-- This module builds on "TaskStore" and adds priority ordering,
-- retry policies and a small typeclass for pluggable clocks.
module Data.Task.Scheduler
  ( Priority(..)
  , RetryPolicy(..)
  , Clock(..)
  , Slot
  , schedule
  , nextSlot
  , describePolicy
  ) where

import Data.List (sortOn)
import Data.Maybe (fromMaybe, mapMaybe)
import qualified Data.Map.Strict as Map
import Control.Monad.State (State, get, put)

-- | How urgently a task should run.
data Priority
  = Low
  | Normal
  | High
  deriving (Show, Eq, Ord)

-- | Retry behaviour attached to a scheduled task.
data RetryPolicy = RetryPolicy
  { retryLimit   :: Int
  , retryBackoff :: Double
  } deriving (Show, Eq)

-- | A slot is a millisecond offset from the epoch.
newtype Slot = Slot Int
  deriving (Show, Eq, Ord)

-- | Convenient alias for a priority-indexed bucket of names.
type Buckets = Map.Map Priority [String]

-- | Anything that can report the current time.
class Clock c where
  -- | Current time, in milliseconds.
  now :: c -> Int
  -- | Sleep for the given number of milliseconds.
  sleepFor :: c -> Int -> Int

-- | Orders tasks so that high priority names come first.
--
-- Ties keep their original relative order, because 'sortOn' is stable.
schedule :: [(String, Priority)] -> [String]
schedule = map fst . sortOn (negate . fromEnum . snd)

-- | Computes the slot following @s@, spaced by the policy backoff.
nextSlot :: RetryPolicy -> Slot -> Slot
nextSlot policy (Slot s) =
  let step = ceiling (retryBackoff policy * 1000.0)
  in Slot (s + max 1 step)

-- | Renders a policy as a human readable string.
--
-- The escape sequences below exercise the string lexer: "quote \" and
-- backslash \\ inside a literal", plus a char literal '\n'.
describePolicy :: RetryPolicy -> String
describePolicy p =
  "retry up to " ++ show (retryLimit p) ++ " times, backoff \"" ++ show (retryBackoff p) ++ "\"" ++ ['\n']

-- | Groups names into buckets keyed by priority.
bucketize :: [(String, Priority)] -> Buckets
bucketize = foldr insert Map.empty
  where
    insert (name, prio) acc = Map.insertWith (++) prio [name] acc

{- | A block-form Haddock comment attached to a helper.

   It spans several lines and contains a -- that must not end it.
-}
pendingNames :: Buckets -> [String]
pendingNames buckets = fromMaybe [] (Map.lookup High buckets)
